librz/analysis/arch/avr/avr_il.c

```diff
@@ -0,0 +1,3159 @@
+}
+
+static RzILOpEffect *avr_il_brlt(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if N ^ V = 1
+	ut16 k = aop->param[0];
+
+	RzILOpBool *N = VARG(AVR_SREG_N);
+	RzILOpBool *V = VARG(AVR_SREG_V);
+	RzILOpPure *when = XOR(N, V);
+	return avr_il_branch_when(aop, analysis, k, when, true);
+}
+
+static RzILOpEffect *avr_il_brmi(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if N = 1
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_N);
+	return avr_il_branch_when(aop, analysis, k, when, true);
+}
+
+static RzILOpEffect *avr_il_brne(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if Z = 0
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_Z);
+	return avr_il_branch_when(aop, analysis, k, when, false);
+}
+
+static RzILOpEffect *avr_il_brpl(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if N = 0
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_N);
+	return avr_il_branch_when(aop, analysis, k, when, false);
+}
+
+static RzILOpEffect *avr_il_brtc(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if T = 0
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_T);
+	return avr_il_branch_when(aop, analysis, k, when, false);
+}
+
+static RzILOpEffect *avr_il_brts(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if T = 1
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_T);
+	return avr_il_branch_when(aop, analysis, k, when, true);
+}
+
+static RzILOpEffect *avr_il_brvc(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if V = 0
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_V);
+	return avr_il_branch_when(aop, analysis, k, when, false);
+}
+
+static RzILOpEffect *avr_il_brvs(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// branch if V = 1
+	ut16 k = aop->param[0];
+
+	RzILOpBool *when = VARG(AVR_SREG_V);
+	return avr_il_branch_when(aop, analysis, k, when, true);
+}
+
+static RzILOpEffect *avr_il_bst(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// Stores bit b from Rd to the T Flag in SREG (Status Register)
+	ut16 Rd = aop->param[0];
+	ut16 b = aop->param[1];
+	avr_return_val_if_invalid_gpr(Rd, NULL);
+
+	RzILOpPure *reg, *bit;
+
+	reg = AVR_REG(Rd);
+	bit = AVR_IMM(1u << b);
+	bit = LOGAND(reg, bit);
+	bit = NON_ZERO(bit);
+	return SETG(AVR_SREG_T, bit);
+}
+
+static RzILOpEffect *avr_il_call(AVROp *aop, AVROp *next_op, ut64 pc, RzAnalysis *analysis) {
+	// PC = k
+	ut32 k = aop->param[0];
+	k <<= 16;
+	k |= aop->param[1];
+
+	RzILOpPure *val, *num;
+	RzILOpEffect *jmp, *push, *sub;
+
+	jmp = avr_il_jump_relative(aop, analysis, k);
+
+	num = AVR_PC(pc);
+	val = VARG(AVR_SP);
+	val = AVR_ADDR(val);
+	push = STOREW(val, num);
+
+	num = AVR_IMM16(AVR_ADDR_SIZE / 8);
+	val = VARG(AVR_SP);
+	val = SUB(val, num);
+	sub = SETG(AVR_SP, val);
+
+	return SEQ3(push, sub, jmp);
```

Comment:
As a general comment, I think it would be more readable to just write the code like:
```c
return SEQ3(
	STOREW(AVR_ADDR(VARG(AVR_SP)), AVR_PC(pc)),
	SETG(AVR_SP, SUB(VARG(AVR_SP), AVR_IMM16(AVR_ADDR_SIZE / 8))),
	avr_il_jump_relative(aop, analysis, k));
```
In particular since now you don't have to write out all the long `rz_il_op_new_...` calls anymore.
But of course it is a matter of taste too.

Reply:
i disagree. i wrote this in this way because some instructions were insanely complex to implement